add inter-component communication vision document (#116)
Apodini/Apodini: Documentation/Components/Inter-Component Communication.md

![document type: vision](https://apodini.github.io/resources/markdown-labels/document_type_vision.svg)

# Inter-Component Communication

## Summary

- The APIs described in this document are used to access, from within a handler's `handle()` function, the functionality provided by another handler

> - The `InvocableHandler` protocol, which interits from `IdentifiableHandler`, is used to denote `Handler`s which can be remotely invoked (ie, from within a `Handler.handle()` function)

Reviewer: ```suggestion
- The `InvocableHandler` protocol, which inherits from `IdentifiableHandler`, is used to denote `Handler`s which can be remotely invoked (ie, from within a `Handler.handle()` function)
```

- The `RemoteHandlerInvocationManager` type implements a helper object which is responsible for coordinating inter-component communication
- This helper object (you can think of this as a kind of mediator) has knowledge of the web service's structure and use that to determine how to realise interactions between handlers
- The API as currenly defined only supports invoking components which use the request-response communication pattern

## The `InvocableHandler` protocol

```swift
public protocol InvocableHandler: IdentifiableHandler where Response: Decodable {
    typealias ParametersStorageProtocol = InvocableHandlerParametersStorageProtocol
    associatedtype ParametersStorage: ParametersStorageProtocol = _EmptyParametersStorage<Self> where ParametersStorage.HandlerType == Self
}
```

> An `InvocableHandler` is an `IdentifiableHandler` whose `Response` type conforms to `Decodable`, and which optionally defines a custom type for passing parameters to a remote invcation (see below for an explanation of how parameter passsing works).

Reviewer: ```suggestion
An `InvocableHandler` is an `IdentifiableHandler` whose `Response` type conforms to `Decodable`, and which optionally defines a custom type for passing parameters to a remote invocation (see below for an explanation of how parameter passsing works).
```

## Defining and using invocable handlers

Simple example of a handler which accesses another handler's functionality.

```swift
struct RandomNumberGenerator: InvocableHandler {
    class HandlerIdentifier: ScopedHandlerIdentifier<RandomNumberGenerator> {
        static let main = HandlerIdentifier("main")
    }
    let handlerId = HandlerIdentifier.main
    
    @Parameter var lowerBound: Int
    @Parameter var upperBound: Int
    
    func handle() -> Int {
        Int.ramdom(in: lowerBound..<upperBound)
    }
}

struct Greeter: Handler {
    private var RHI = RemoteHandlerInvocationManager()
    
    @Parameter var name: String
    
    func handle() -> EventLoopFuture<String> {
        RHI.invoke(
            RandomNumberGenerator.self,
            identifiedBy: .main,
            parameters: [
                .init(\.$lowerBound, 0),
                .init(\.$upperBound, name.length)
            ]
        ).map { number in
            "Hello, \(name)! Your lucky number is \(number)."
        }
    }
}
```

## The Remote Invocation Manager

> The invocation manager acts as a mediator overseeing all inter-handler interactions for a handler. Each handler which wants to other handlers defines its own instance of the invocation manager.

Reviewer: ```suggestion
The invocation manager acts as a mediator overseeing all inter-handler interactions for a handler. Each handler which wants to interact with other handlers defines its own instance of the invocation manager.
```

### What's an interaction?

> An interaction is an event where date are transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.

Reviewer: ```suggestion
An interaction is an event where data is transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.
```

I would use "data is" but "data are" is apparently also correct: https://www.theguardian.com/news/datablog/2010/jul/16/data-plural-singular

Broadly speaking, we differentiate between two kinds of interactions:

- Local (in-process) interactions
  - These are interactions where sender and target are running in the same process
  - In this case, the interaction is dispatched locally, ie the target endpoint's `handle()` function executed on the current event loop
- Remote (out-of-process) interactions
  - These are interactions where sender and target are running in different processed (potentially even on different machines and different networks)
  - Remote interactions are available when running a web service as a distributed system (ie, a system of distinct nodes, each of which implement a subset of the web service's exported endpoints)
  - In this case, the interaction is realised by finding the correct node within our distributed system, invoking the target component there, and returning the result back to the sender

### What does the invocation manager do?

The invocation manager is responsible for deciding whether the target component should be invoked locally or remotely, thus acting as a dispatcher. It is also responsible for realising the interaction.

> The invocation manager has access to the full structure of the web service (eg via the `WebServiceModel` object) and the deployment structure structure (ie information about the number of processes running the web service, the endpoints implemented by each process, etc).

Reviewer: ```suggestion
The invocation manager has access to the full structure of the web service (eg via the `WebServiceModel` object) and the deployment structure (ie information about the number of processes running the web service, the endpoints implemented by each process, etc).
```

```

Where `CollectedParameter` is a struct storing:

- a key path to an `@Parameter` within an invocable handler
- the value which should be passed for this parameter

(`_EmptyParametersStorage ` is a non-initialisable type which a Handler can use to opt in to the array-based parameter passing. For `InvocableHandler`s which do not specify a `ParametersStorage` type, the parameter storage defaults to this type.)

Similar to how the `ParametersStorage` consisted of the parameter values alongside an array mapping the key path to a value (within the parameters storage) to the key path of its respective `@Parameter` (within the Handler), the `CollectedParameter`s expected by this function consist of the parameter value and the key path to the handler's `@Parameter`.

Entries for `@Parameter`s which define a default value may be omitted from the parameters array.  
If the parameters array contains multiple entries for the same `@Parameter` key path, the last one is used.  
Incomplete parameter lists (eg missing parameters) result in a run-time error.

The advantage of this parameter-passing approach is that the handler doesn't have to write boilerplate code.  

> The downside however is the loss of compile-time well-formedness checking.

Reviewer: ```suggestion
The downside, however, is the loss of compile-time well-formedness checking.
```
